### PR TITLE
Enable batch without attention_mask

### PR DESCRIPTION
**Pull request by michal2409**
# What does this PR do ?

1. Allow for empty attention_mask in the input batch

# Changelog

- Add specific line by line info of high level changes in this PR.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Review by yaoyu-33 — DISMISSED** [review]

File: src/megatron/bridge/training/eval.py [hunk 1]
```diff
@@ -137,6 +137,9 @@ def evaluate(
                 micro_batch_size=state.cfg.train.micro_batch_size,
                 forward_only=True,
             )
+            # Synchronize the full_iteration cuda graph to avoid race conditions
+            if state.cfg.model.cuda_graph_impl == "local" and "full_iteration" in state.cfg.model.cuda_graph_scope:
```
Comment (vasunvidia) [inline src/megatron/bridge/training/eval.py]: This needs to be 
> CudaGraphScope.full_iteration in state.cfg.model.cuda_graph_scope
Please refer to https://github.com/NVIDIA/Megatron-LM/pull/2657/files

File: tests/unit_tests/training/test_eval.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+        mock_wrapper_instance = Mock(return_value=[{}])
+        mock_cuda_graph_wrapper.return_value = mock_wrapper_instance
+
+        mock_prepare_forward_step.return_value = Mock()
+
+        # Create state with CUDA graph config
+        state = self._create_mock_state(
+            cuda_graph_impl="local",
+            cuda_graph_scope=["full_iteration"],
```
Comment (vasunvidia) [inline tests/unit_tests/training/test_eval.py]: CudaGraphScope.full_iteration

File: tests/unit_tests/training/test_eval.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+        mock_fwd_bwd = Mock(return_value=[{}])
+        mock_get_fwd_bwd_func.return_value = mock_fwd_bwd
+
+        mock_prepare_forward_step.return_value = Mock()
+
+        # Create state WITHOUT cuda graph enabled
+        state = self._create_mock_state(
+            cuda_graph_impl="none",
+            cuda_graph_scope=["full_iteration"],
```
Comment (vasunvidia) [inline tests/unit_tests/training/test_eval.py]: CudaGraphScope.full_iteration

File: tests/unit_tests/training/test_eval.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+        mock_fwd_bwd = Mock(return_value=[{}])
+        mock_get_fwd_bwd_func.return_value = mock_fwd_bwd
+
+        mock_prepare_forward_step.return_value = Mock()
+
+        # Create state with local impl but NOT full_iteration scope
+        state = self._create_mock_state(
+            cuda_graph_impl="local",
+            cuda_graph_scope=["attn"],  # Not full_iteration
```
Comment (vasunvidia) [inline tests/unit_tests/training/test_eval.py]: CudaGraphScope.attn

File: src/megatron/bridge/training/gpt_step.py [hunk 5]
```diff
@@ -125,7 +125,7 @@ def get_batch(
         batch["tokens"],
         batch["labels"],
         batch["loss_mask"],
-        batch["attention_mask"],
+        batch.get("attention_mask"),
```
Comment (gautham-kollu) [inline src/megatron/bridge/training/gpt_step.py]: np: Add a comment saying. Attention_mask is optional for pre-training as a casual mask is generated automatically.
